+nosnoc/+discrete_time_problem/Cls.m

```diff
@@ -0,0 +1,1600 @@
+                            obj.g.equidistant_control_grid(ii) = {g_eq_grid, -inf, 0};
+                            obj.f = obj.f + opts.rho_terminal_numerical_time*obj.w.s_numerical_time(ii);
+                        else
+                            obj.g.equidistant_control_grid(ii) = {t_stage-sum_h};
+                        end
+                    end
+                end
+            end
+
+            x_end = obj.w.x(opts.N_stages,opts.N_finite_elements(opts.N_stages),opts.n_s+rbp);
+            z_end = obj.w.z(opts.N_stages,opts.N_finite_elements(opts.N_stages),opts.n_s+rbp);
+            % Terminal cost
+            obj.f = obj.f + dcs.f_q_T_fun(x_end, z_end, v_global, p_global);
+
+            % Terminal_lsq_cost
+            if ~isempty(model.x_ref_end_val)
+                obj.f = obj.f + h0*opts.N_finite_elements(ii)*dcs.f_lsq_T_fun(x_end,...
+                    model.x_ref_end_val,...
+                    p_global);
+            end
+
+            % Terminal constraint
+            if opts.relax_terminal_constraint_homotopy
+                error("Currently unsupported")
+            end
+            g_terminal = dcs.g_terminal_fun(x_end, z_end, v_global, p_global);
+            switch opts.relax_terminal_constraint
+              case ConstraintRelaxationMode.NONE % hard constraint
+                if opts.relax_terminal_constraint_from_above
+                    obj.g.terminal = {g_terminal, model.lbg_terminal, inf*ones(dims.n_g_terminal,1)};
+                else
+                    obj.g.terminal = {g_terminal, model.lbg_terminal, model.ubg_terminal};
+                end
+              case ConstraintRelaxationMode.ELL_1 % l_1
+                obj.w.s_terminal_ell_1 = {{'s_terminal_ell_1', dims.n_g_terminal}, 0, inf, 10};
+
+                g_terminal = [g_terminal-model.lbg_terminal-obj.w.s_terminal_ell_1();
+                    -(g_terminal-model.ubg_terminal)-obj.w.s_terminal_ell_1()];
+                obj.g.terminal = {g_terminal, -inf, 0};
+                obj.f = obj.f + obj.p.rho_terminal_p()*sum(obj.w.s_terminal_ell_1());
+              case ConstraintRelaxationMode.ELL_2 % l_2
+                     % TODO(@anton): this is as it was implemented before. should handle lb != ub?
+                obj.f = obj.f + obj.p.rho_terminal_p()*(g_terminal-model.lbg_terminal)'*(g_terminal-model.lbg_terminal);
+              case ConstraintRelaxationMode.ELL_INF % l_inf
+                obj.w.s_terminal_ell_inf = {{'s_terminal_ell_inf', 1}, 0, inf, 1e3};
+
+                g_terminal = [g_terminal-model.lbg_terminal-obj.w.s_terminal_ell_inf();
+                    -(g_terminal-model.ubg_terminal)-obj.w.s_terminal_ell_inf()];
+                obj.g.terminal = {g_terminal, -inf, 0};
+                obj.f = obj.f + obj.p.rho_terminal_p()*obj.w.s_terminal_ell_inf();
+            end
+        end
+
+        % TODO(@anton) is there a way to make this less horribly ugly?
```

Comment:
Maybe we should not judge too much in the public code. :P

Reply:
Eh, I will drop this but considering the similar comments that exist in other projects (IPOPT, etc  just in our field) I wouldn't worry too much about it.